extensions/oplogPopulator/OplogPopulator.js

```diff
@@ -27,80 +32,388 @@ class OplogPopulator {
+            // rule is enabled
+            extFilter[`value.${field}.rules`] = {
+                $elemMatch: {
+                    ruleStatus: 'Enabled',
+                },
+            };
+            filter.$or.push(extFilter);
+        }
+        // ingestion filter
+        if (this._activeExtensions.includes('ingestion')) {
+            const field = constants.extensionConfigField.ingestion;
+            const extFilter = {};
+            extFilter[`value.${field}.status`] = 'enabled';
+            filter.$or.push(extFilter);
+        }
+        try {
+            const buckets = await this._metastore.find(filter)
+                .project({ _id: 1 })
+                .map(bucket => bucket._id)
+                .toArray();
+            return buckets;
+        } catch (err) {
+            this._logger.error('Error querying buckets from MongoDB', {
+                method: 'OplogPopulator._getBackbeatEnabledBuckets',
+                error: err.message,
+            });
+            throw errors.InternalError.customizeDescription(err.message);
+        }
+    }
+
+    /**
+     * Makes new connector pipeline to listen to
+     * specified buckets
+     * @param {string[]} bucketNames list of buckets to listen to
+     * @returns {Object} new connector pipeline
+     */
+    _generateNewConnectorPipeline(bucketNames) {
+        const pipeline = [
+            {
+                $match: {
+                    'ns.coll': {
+                        $in: bucketNames,
+                    }
+                }
+            }
+        ];
+        return JSON.stringify(pipeline);
+    }
+
+    /**
+     * start listening to bucket
+     * @param {string} bucketName bucket name
+     * @returns {undefined}
+     */
+    _listenToBucket(bucketName) {
+        this._bucketsForConnector.add(bucketName);
+        this.updateConnectorPipeline(constants.defaultConnectorName, [...this._bucketsForConnector]);
+    }
+
+    /**
+     * stops listening to bucket
+     * @param {string} bucketName bucket name
+     * @returns {undefined}
+     */
+    _stopListeningToBucket(bucketName) {
+        this._bucketsForConnector.delete(bucketName);
+        this.updateConnectorPipeline(constants.defaultConnectorName, [...this._bucketsForConnector]);
+    }
+
+    /**
+     * Check if buckets has at least one backbeat extension active
+     * @param {BucketInfo} bucketMetadata bucket metadata
+     * @returns {boolean} is bucket backbeat enabled
+     */
+    _isBucketBackbeatEnabled(bucketMetadata) {
+        const areExtensionsEnabled = [];
+        // notification extension
+        if (this._activeExtensions.includes('notification')) {
+            const field = constants.extensionConfigField.notification;
+            areExtensionsEnabled.push(Boolean(bucketMetadata[field] !== null));
+        }
+        // replication extension
+        if (this._activeExtensions.includes('replication')) {
+            const field = constants.extensionConfigField.replication;
```

Comment:
same as above, the business logic should be moved out of this (generic) code, somehow ⬆️

Note that this may be done in a separate techdebt ticket, to keep momentum and avoid conflicts/merges (→ please create it, in that case)

Reply:
Techdebt ticket created